Shifty.App/DomainModels/MenuItem.cs

```diff
@@ -0,0 +1,55 @@
+using System;
+using System.Data.Common;
+using Shifty.Api.Generated.AnalogCoreV1;
+using Shifty.Api.Generated.AnalogCoreV2;
+
+namespace Shifty.App.DomainModels
+{
+    public class MenuItem {
+        public int Id { get; init; }
+        public string Name { get; set; }
+        public static MenuItem FromDto(MenuItemResponse dto)
+        {
+            return new MenuItem()
+            {
+                Id = dto.Id,
+                Name = dto.Name
+            };
+        }
+
+        public static AddMenuItemRequest ToAddRequest(MenuItem menuItem)
+        {
+            return new AddMenuItemRequest()
+            {
+                Name = menuItem.Name
+            };
+        }
+
+        public static UpdateMenuItemRequest ToUpdateRequest(MenuItem menuItem)
+        {
+            return new UpdateMenuItemRequest()
+            {
+                Name = menuItem.Name
+            };
+        }
+
+        public override string ToString()
+        {
+            return Name;
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj is not MenuItem)
+            {
+                return false;
+            }
+            return ((MenuItem)obj).Id == Id;
+        }
+
+        public override int GetHashCode()
+        {
+            return HashCode.Combine(Id, Name);
+        }
```

Comment:
@TTA777 comments based on your verbal questions on Tuesday:

These two methods are required as we are using them in a dictionary in the product manager.
Without this, the dictionary is not able to update the current dictionary to match the product currently being edited. 
Another solution is this:
https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.equalitycomparer-1?view=net-8.0

In my opinion it is not as pretty as our current solution - It seems that we would be expected to implement a new class and use that in the constructor for our dictionary, and then have this new class implement the equalitycomparer interface with type MenuItem.

But please let me know if my understanding is wrong/incomplete or if this is best practice.

Reply:
Okay, so now I remember the confusion with the dictionary... I agree that the dictionary i explicitly created did not do anything.

However, somewhere "underneath the hood", a dictionary is used for the implementation of the mudcombobox from the mud extensions. However, I do not have direct access to this dictionary, so I am having difficulties utilizing the EqualityComparer.
Setting it directly on the <MudComboBox> object:
```cs
<MudComboBox
                        T="MenuItem" 
                        Comparer="EqualityComparer<MenuItem>.Create(
                            (mi1, mi2) => mi1.Id == mi2.Id
                        )" 
/>
```
gives the following error:
```
Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Unable to set property 'Comparer' on object of type 'MudExtensions.MudComboBox`1[[Shifty.App.DomainModels.MenuItem, Shifty.App, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]]'. The error was: Value cannot be null. (Parameter 'collection')
System.InvalidOperationException: Unable to set property 'Comparer' on object of type 'MudExtensions.MudComboBox`1[[Shifty.App.DomainModels.MenuItem, Shifty.App, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]]'. The error was: Value cannot be null. (Parameter 'collection')
 ---> System.ArgumentNullException: Value cannot be null. (Parameter 'collection')
   at System.Collections.Generic.HashSet`1[[Shifty.App.DomainModels.MenuItem, Shifty.App, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]]..ctor(IEnumerable`1 collection, IEqualityComparer`1 comparer)
   at MudExtensions.MudComboBox`1[[Shifty.App.DomainModels.MenuItem, Shifty.App, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].set_Comparer(IEqualityComparer`1 value)
   at Microsoft.AspNetCore.Components.Reflection.PropertySetter.CallPropertySetter[MudComboBox`1,IEqualityComparer`1](Action`2 setter, Object target, Object value)
   at Microsoft.AspNetCore.Components.Reflection.PropertySetter.SetValue(Object target, Object value)
   at Microsoft.AspNetCore.Components.Reflection.ComponentProperties.<SetProperties>g__SetProperty|3_0(Object target, PropertySetter writer, String parameterName, Object value)
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Components.Reflection.ComponentProperties.<SetProperties>g__SetProperty|3_0(Object target, PropertySetter writer, String parameterName, Object value)
   at Microsoft.AspNetCore.Components.Reflection.ComponentProperties.SetProperties(ParameterView& parameters, Object target)
   at Microsoft.AspNetCore.Components.ParameterView.SetParameterProperties(Object target)
   at Microsoft.AspNetCore.Components.ComponentBase.SetParametersAsync(ParameterView parameters)
   at MudExtensions.MudBaseInputExtended`1[[Shifty.App.DomainModels.MenuItem, Shifty.App, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].<>n__1(ParameterView parameters)
   at MudExtensions.MudBaseInputExtended`1.<SetParametersAsync>d__212[[Shifty.App.DomainModels.MenuItem, Shifty.App, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()
```

Regardless, seems like some language support broke for my visual studio between 8.0.1 (used in analog-core) and 8.0.2 (used in shifty now), so I am also having difficulties attempting other solutions. 

Without the `Equals` and `ToHashCode` methods on the MenuItem class, it is not possible to determine how to map a new items `EligibleMenuItems` to the checkbox list generated in the `EditTemplate` (and the other way around upon save).

The documentation for the [MudComboBox](https://codebeam-mudextensions.pages.dev/mudcombobox) does not state anything in relation to the comparer or the equals requirement so that is why I am also a bit uncertain on how to properly implement this.

I have found, based on the [documentation](https://mudblazor.com/components/autocomplete#strict-mode) from the `MudAutoComplete` (which the `MudComboBox` attempts to use, mixed with the `MudSelect`), that we can use a record type instead of a class. See my latest commit for using this. As far as I can see, this works as expected however I am uncertain if this will introduce issues for the [Menu Item Management](https://github.com/AnalogIO/shifty-webapp/tree/feature/frepe/menu-item-management).

Reply:
I have looked a little bit more. I can't see any EqualityComparer or Comparer object referenced in the documentation, however I found this code in their MudComboBox which makes it seems like, it should be there.
https://github.com/CodeBeamOrg/CodeBeam.MudBlazor.Extensions/blob/1c05d6cf7b95c7c0d10e12ec01755756f0841199/CodeBeam.MudBlazor.Extensions/Components/ComboBox/MudComboBox.razor.cs#L536

I wanted to open an issue on their github repository with a minimum example and ask if it is a bug or if there is some documentation somewhere, however I was unable to construct a minimum example through their online editor in reasonable time. I might come back to this later, but I don't think we can wait until that is fixed anyway. I have pushed the new API json so everything works again.